src/components/input/SelectNext.tsx

```diff
@@ -0,0 +1,195 @@
+  onChange: (newValue: T) => void;
+  label: ComponentChildren;
+  disabled?: boolean;
+};
+
+function SelectMain<T>({
+  label,
+  value,
+  onChange,
+  children,
+  disabled,
+  classes,
+  elementRef,
+}: SelectProps<T>) {
+  const [isListboxOpen, setIsListboxOpen] = useState(false);
+  const [shouldListboxDropUp, setShouldListboxDropUp] = useState(false);
+  const closeListbox = useCallback(() => setIsListboxOpen(false), []);
+  const wrapperRef = useRef<HTMLDivElement | null>(null);
+  const listboxRef = useRef<HTMLDivElement | null>(null);
+  const listboxId = useId();
+  const buttonRef = useSyncedRef(elementRef);
+  const buttonId = useId();
+
+  const selectValue = useCallback(
+    (newValue: unknown) => {
+      onChange(newValue as T);
+      closeListbox();
+    },
+    [closeListbox, onChange],
+  );
+
+  // When clicking away or pressing `Esc`, close the listbox
+  useClickAway(wrapperRef, closeListbox);
+  useKeyPress(['Escape'], closeListbox);
+
+  // Vertical arrow key for options in the listbox
+  useArrowKeyNavigation(wrapperRef, { horizontal: false });
+
+  useLayoutEffect(() => {
+    if (!isListboxOpen) {
+      // Focus button after closing listbox
+      buttonRef.current!.focus();
+      // Reset shouldDropUp so that it does not affect calculations next time
+      // it opens
+      setShouldListboxDropUp(false);
+    } else {
+      const viewportHeight = window.innerHeight;
+      const { top: buttonDistanceToTop, bottom: buttonBottom } =
+        buttonRef.current!.getBoundingClientRect();
+      const buttonDistanceToBottom = viewportHeight - buttonBottom;
+      const { bottom: listboxBottom } =
+        listboxRef.current!.getBoundingClientRect();
+      const listboxDistanceToBottom = viewportHeight - listboxBottom;
+
+      // The listbox should drop up only if there's not enough space below it for
+      // the listbox max height, and there's also more space above
+      setShouldListboxDropUp(
+        listboxDistanceToBottom < 0 &&
+          buttonDistanceToTop > buttonDistanceToBottom,
+      );
+    }
+  }, [buttonRef, isListboxOpen]);
+
+  return (
+    <div className="relative" ref={wrapperRef}>
+      <Button
+        id={buttonId}
+        variant="custom"
+        classes={classnames(
+          'w-full flex border',
+          'bg-grey-0 disabled:bg-grey-1 disabled:text-grey-6',
+          classes,
+        )}
+        expanded={isListboxOpen}
+        pressed={isListboxOpen}
+        disabled={disabled}
+        aria-haspopup="listbox"
+        aria-controls={listboxId}
+        elementRef={buttonRef}
+        onClick={() => setIsListboxOpen(prev => !prev)}
+        onKeyDown={e => {
+          if (e.key === 'ArrowDown' && !isListboxOpen) {
+            setIsListboxOpen(true);
+          }
+        }}
+        data-testid="select-toggle-button"
+      >
+        {label}
+        <div className="grow" />
+        {isListboxOpen ? <MenuCollapseIcon /> : <MenuExpandIcon />}
+      </Button>
```

Comment:
As noted in comments, this button isn't identified as controlling a popup listbox in VoiceOver at least (not sure about other screen readers). Marking the button as being `role=combobox` resolved this.

Reply:
The button was identified as a toggle becase I added `aria-pressed` attribute by mistake. I'll remove it.

I think I'm not going to go in the combobox direction for now, because then a11y rules require it to have a label or aria-label, whereas its own content is enough for screen readers while it has role button.

I considered inferring the `aria-label` attribute from the toggle content, but that would require a plain-text version of it (it allows ComponentChildren at the moment).

I'm not completely sure what's the right approach, but I see most UI libraries not using the combobox approach.

* https://mui.com/material-ui/react-select/
* https://headlessui.com/react/listbox
* https://getbootstrap.com/docs/5.3/components/dropdowns/#single-button